appium/webdriver/webdriver.py

```diff
@@ -142,6 +148,24 @@ def __init__(self, command_executor='http://127.0.0.1:4444/wd/hub',
         By.IMAGE = MobileBy.IMAGE
         By.CUSTOM = MobileBy.CUSTOM
 
+    def _update_command_executor(self, keep_alive):
+        """Update command executor following directConnect feature"""
+        protocol = self.capabilities['directConnectProtocol']
+        hostname = self.capabilities['directConnectHost']
+        port = self.capabilities['directConnectPort']
+        path = self.capabilities['directConnectPath']
+        executor = '{scheme}://{hostname}:{port}{path}'.format(
+            scheme=protocol,
+            hostname=hostname,
+            port=port,
+            path=path
+        )
+
+        logger.info('Update request endpoint to %s', executor)
+        # Override command executor
+        self.command_executor = RemoteConnection(executor, keep_alive=keep_alive)
+        self._addCommands()
```

Comment:
do we want to execute _addCommands twice?

Reply:
Yes.
`self.command_executor = RemoteConnection(executor, keep_alive=keep_alive)` replaces executor with new instance. The new one does not have commands added by `_addCommands()` ( https://github.com/appium/python-client/pull/338/files#diff-d2ce7e5cde58c87021c848f4b40b2106R216 fails if we do not add this line. )